### DISCUSSION TITLE
generating relocatable code

### DISCUSSION THREAD

**Discussion by iacore** [vnmakarov/mir#380] [General]
Is it possible, for example, to generate relocatable x86_64 code and put it in ELF?

I think the MIR JIT compiler puts the code in memory to be executable, but I don't know how can I save it to disk and use it in another process (after the system has restarted).

**Comment by vnmakarov** [comment]
> Is it possible, for example, to generate relocatable x86_64 code and put it in ELF?

Not today.  First we need to implement position-independent code (because when you put the code into another process you probably can not guarantee the same start address).  Implementing it would take considerable work, especially as it should be done for 5 targets.  On the other hand, it is partially implemented that even some specific code generated by MIR can be already PIC.

People already proposed to generate PIC an ELF.  It is on my list (only I would like to know when I can start to work on this list :(. 

> I think the MIR JIT compiler puts the code in memory to be executable, but I don't know how can I save it to disk and use it in another process (after the system has restarted).

I think implementing PIC code would the first important step to achieve this.  ELF would be overkill for this task.  We could implement simpler target specific relocation mechanism for external entities.

On the other hand, the code could be re-generated from MIR-text/binary/C and you will not need all these new features.  I guess you have too much code and are not satisfied with the generation speed.

**Reply by iacore** [reply]
>  I guess you have too much code and are not satisfied with the generation speed.

No, not at all. I was trying to explore what MIR can do.

**Reply by rsepassi** [reply]
Would it be easier to target relocatable first before position-independent? i.e. Be able to generate object files that could be passed to a linker (ld, lld) (or maybe even the MIR linker?).
